corehq/apps/data_interfaces/tasks.py

```diff
@@ -111,3 +113,18 @@ def run_case_update_rules_for_domain(domain, now=None):
         for rule in rules:
             rule.last_run = now
             rule.save()
+
+
+@task(queue='background_queue', acks_late=True, ignore_result=True)
+def run_case_update_rules_on_save(domain, case_type, case_id):
+    rules = AutomaticUpdateRule.by_domain(domain).filter(case_type=case_type)
+    case = CaseAccessors(domain).get_case(case_id)
+    now = datetime.utcnow()
+    key = 'case-update-on-save-case-{case}'.format(case=case_id)
+    with CriticalSection([key]):
```

Comment:
I had to keep the critical section after making sure the case hadnt been updated by this task, otherwise the second instantiation was hitting against the critical section until timing out.